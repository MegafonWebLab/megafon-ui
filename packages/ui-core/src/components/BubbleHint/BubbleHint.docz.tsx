import * as React from 'react';

export const BubbleHintWrapper = ({ children }) => {
    return (
        <div style={{ height: '400px', display: 'flex', 'justifyContent': 'center', 'alignItems': 'center' }}>
            {children}
        </div>
    );
};
