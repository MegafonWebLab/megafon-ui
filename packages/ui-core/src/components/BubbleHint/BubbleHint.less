@import '~styles/base.less';
@b: e('.mfui-bubble-hint');

@{b} {
    .fontFamily();
    position: relative;
    text-align: left;

    &__trigger {
        cursor: pointer;

        svg {
            // fix for edge (svg elements call mouseleave event)
            pointer-events: none;
        }
    }
    &__popup {
        visibility: hidden;
        z-index: 10;
    }

    &__popup_width_small {
        width: 216px;
    }
    &__popup_width_normal {
        width: 268px;
    }
    &__popup_width_big {
        width: 330px;
        @media @mobileSM {
            width: 280px;
        }
    }

    &__popup-inner {
        cursor: default;
        position: relative;
        box-sizing: border-box;
        box-shadow: 0 0 20px 2px rgba(0, 0, 0, 0.2);
        background-color: @clearWhite;
    }
    &__popup-inner_padding_normal {
        padding: 28px;
    }
    &__popup-inner_padding_bottom-small {
        padding: 25px 25px 17px 25px;
    }
    &__popup-inner_padding_none {
        padding: 0;
    }

    &__popup-inner_align_left {
        text-align: left;
    }
    &__popup-inner_align_center {
        text-align: center;
    }
    &__popup-inner_align_right {
        text-align: right;
    }

    &__content {
        .commonFont();
        color: @freshAsphalt;
    }

    [data-placement='right'] {
        padding-left: 15px;
    }
    [data-placement='left'] {
        padding-right: 15px;
    }
    [data-placement='top'] {
        padding-bottom: 15px;
    }
    [data-placement='bottom'] {
        padding-top: 15px;
    }
    &__popup_show {
        visibility: visible;
    }

    &__arrow {
        position: absolute;
        overflow: hidden;

        &:before {
            box-sizing: border-box;
            position: absolute;
            content: '';
            width: 16px;
            height: 16px;
            background-color: @white;
            box-shadow: 0 0 4px rgba(0, 0, 0, 0.1);
            transform: rotate(45deg);
        }
    }
    [data-placement='right'] &__arrow {
        right: 100%;
        left: auto !important;
        width: 10px;
        height: 26px;
        &:before {
            top: 3px;
            left: 6px;
        }
    }
    [data-placement='left'] &__arrow {
        left: 100% !important;
        width: 10px;
        height: 26px;
        &:before {
            top: 3px;
            right: 6px;
        }
    }
    [data-placement='top'] &__arrow {
        top: 100% !important;
        width: 26px;
        height: 10px;
        &:before {
            bottom: 6px;
            right: 5px;
        }
    }
    [data-placement='bottom'] &__arrow {
        bottom: 100%;
        top: auto !important;
        width: 26px;
        height: 10px;
        &:before {
            top: 6px;
            right: 5px;
        }
    }

}
