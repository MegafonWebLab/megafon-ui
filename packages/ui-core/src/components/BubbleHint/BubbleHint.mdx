---
name: BubbleHint
menu: Components
route: components/bubble-hint
---

import { Playground, Props } from 'docz';
import BubbleHint from './BubbleHint';
import Paragraph from '../Paragraph/Paragraph';
import Button from '../Button/Button';
import { BubbleHintWrapper } from './BubbleHint.docz';

# BubbleHint

## Properties

<Props of={BubbleHint} />

## Basic usage

<Playground>
    <BubbleHintWrapper>
        <BubbleHint
            placement="top"
            trigger={<Button>BubbleHint</Button>}
        >
            <Paragraph hAlign="center" sizeAll="small">
                Please be careful when filling out the forms
            </Paragraph>
        </BubbleHint>
    </BubbleHintWrapper>
</Playground>

<Playground>
    <BubbleHintWrapper>
        <BubbleHint
            popupAlign="left"
            placement="right"
            trigger={<Button>BubbleHint</Button>}
        >
            <Paragraph hAlign="center" sizeAll="small">
                Please be careful when filling out the forms
            </Paragraph>
        </BubbleHint>
    </BubbleHintWrapper>
</Playground>

<Playground>
    <BubbleHintWrapper>
        <BubbleHint
            placement="bottom"
            trigger={<Button>BubbleHint</Button>}
        >
            <Paragraph hAlign="center" sizeAll="small">
                Please be careful when filling out the forms
            </Paragraph>
        </BubbleHint>
    </BubbleHintWrapper>
</Playground>

<Playground>
    <BubbleHintWrapper>
        <BubbleHint
            placement="left"
            trigger={<Button>BubbleHint</Button>}
            onClick={() => alert('Clicked!')}
        >
            <Paragraph hAlign="center" sizeAll="small">
                Please be careful when filling out the forms
            </Paragraph>
        </BubbleHint>
    </BubbleHintWrapper>
</Playground>
