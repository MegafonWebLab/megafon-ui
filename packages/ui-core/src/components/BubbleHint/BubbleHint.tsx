import * as React from 'react';
import * as PropTypes from 'prop-types';
import './BubbleHint.less';
import { Manager, Popper, Target, Arrow, IPopperProps } from 'react-popper';
import cnCreate from 'utils/cnCreate';
import detectTouch from 'utils/detectTouch';

interface IBubbleHintProps {
    /** Custom class name */
    className?: string;
    /** Padding in popup */
    popupPadding: 'normal' | 'bottom-small' | 'none';
    /** Align in popup */
    popupAlign: 'left' | 'center' | 'right';
    /** Width popup */
    popupWidth: 'small' | 'normal' | 'big';
    /** Use margin left */
    marginLeft: boolean;
    /** Activation on click (default hover) */
    click: boolean;
    /** Popup position */
    placement: 'top' | 'right' | 'bottom' | 'left';
    /** Trigger */
    trigger: JSX.Element[] | Element[] | JSX.Element | string | Element;
    children: JSX.Element[] | Element[] | JSX.Element | string | Element;
    /** MouseEnter handler */
    onMouseEnter?: (e: React.SyntheticEvent<EventTarget>) => void;
    /** MouseLeave handler */
    onMouseLeave?: (e: React.SyntheticEvent<EventTarget>) => void;
    /** Click handler */
    onClick?: (e: React.MouseEvent<HTMLElement>) => void;
}

interface IBubbleHintState {
    show: boolean;
}

const cn = cnCreate('mfui-bubble-hint');
class BubbleHint extends React.Component<Partial<IBubbleHintProps>, IBubbleHintState> {
    static propTypes = {
        popupPadding: PropTypes.oneOf(['normal', 'bottom-small', 'none']),
        popupAlign: PropTypes.oneOf(['left', 'center', 'right']),
        popupWidth: PropTypes.oneOf(['small', 'normal', 'big']),
        marginLeft: PropTypes.bool,
        click: PropTypes.bool,
        placement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
        trigger: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.element),
            PropTypes.element,
            PropTypes.string,
        ]),
        children: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.element),
            PropTypes.element,
            PropTypes.string,
        ]),
        onMouseEnter: PropTypes.func,
        onMouseLeave: PropTypes.func,
        onClick: PropTypes.func,
    };

    static defaultProps = {
        popupAlign: 'left',
        popupWidth: 'normal',
        popupPadding: 'normal',
        placement: 'top',
    };

    container: HTMLElement | null = null;
    trigger: HTMLElement | null = null;
    touchClick: string = detectTouch() ? 'touchstart' : 'click';
    isTouch: boolean = detectTouch();

    constructor(props: IBubbleHintProps) {
        super(props);
        this.state = { show: false };
    }

    componentDidMount() {
        document.addEventListener(this.touchClick, this.handleClickOutside);
    }

    componentWillUnmount() {
        document.removeEventListener(this.touchClick, this.handleClickOutside);
    }

    handleMouseEnter = e => {
        this.setState({ show: true });
        this.props.onMouseEnter && this.props.onMouseEnter(e);
    }

    handleMouseLeave = (e: React.SyntheticEvent<EventTarget>): void => {
        this.setState({ show: false });
        this.props.onMouseLeave && this.props.onMouseLeave(e);
    }

    handleClick = (e: React.MouseEvent<HTMLElement>): void | boolean => {
        if (this.trigger && e.target && this.trigger.contains(e.target as HTMLElement)) {
            this.setState({ show: !this.state.show });
            this.props.onClick && this.props.onClick(e);
        }

        return false;
    }

    handleClickOutside = (e: React.SyntheticEvent<EventTarget> | Event): void | boolean => {
        if ((this.container && this.container.contains(e.target as HTMLElement)) || !this.state.show) {
            return;
        }

        this.setState({ show: false });
    }

    closePopup(): void {
        this.setState({ show: false });
    }

    getHandlers(): {} {
        if (this.isTouch || this.props.click) {
            return {
                onClick: this.handleClick,
            };
        }

        return {
            onMouseEnter: this.handleMouseEnter,
            onMouseLeave: this.handleMouseLeave,
        };
    }

    getContainer = (elem: HTMLDivElement): void => {
        this.container = elem;
    }

    getTrigger = (elem: HTMLDivElement): void => {
        this.trigger = elem;
    }

    render() {
        const popupClasses = { show: this.state.show, width: this.props.popupWidth };
        const popupOptions: IPopperProps = {
            flip: {
                behavior: ['right', 'left', 'bottom', 'top'],
            },
        };

        return (
            <Manager className={cn('', { 'margin-left': this.props.marginLeft }, this.props.className)}>
                <div className={cn('container')}
                    {...this.getHandlers()}
                    ref={this.getContainer}
                >
                    <Target className={cn('target')}>
                        <div
                            className={cn('trigger')}
                            ref={this.getTrigger}
                        >
                            {this.props.trigger}
                        </div>
                    </Target>
                    <Popper eventsEnabled={false}
                        className={cn('popup', popupClasses)}
                        modifiers={popupOptions}
                        placement={this.props.placement}
                    >
                        <div
                            className={cn('popup-inner', {
                                padding: this.props.popupPadding,
                                align: this.props.popupAlign,
                            })}
                        >
                            <div className={cn('content')}>
                                {this.props.children}
                            </div>
                            <Arrow className={cn('arrow')} />
                        </div>
                    </Popper>
                </div>
            </Manager>
        );
    }
}

export default BubbleHint;
