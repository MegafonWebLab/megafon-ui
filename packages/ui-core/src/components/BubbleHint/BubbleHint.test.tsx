import * as React from 'react';
import { shallow } from 'enzyme';
import BubbleHint from './BubbleHint';

const props = {
    click: true,
};

const div = document.createElement('div');

describe('<BubbleHint />', () => {

    beforeEach(() => {
        div.contains = () => { return false; };
        jest.resetModules();
    });

    it('it renders BubbleHint', () => {
        jest.doMock('../../utils/detectTouch', () => {
            return jest.fn(() => true);
        });

        // tslint:disable-next-line:no-shadowed-variable
        const BubbleHint = require('./BubbleHint').default;
        const wrapper = shallow(<BubbleHint />);
        const instance = wrapper.instance() as BubbleHint;
        instance.getTrigger({} as HTMLDivElement);
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders with click', () => {
        const wrapper = shallow(<BubbleHint {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('it checks popup is closed', () => {
        const wrapper = shallow(<BubbleHint />);
        const instance = wrapper.instance() as BubbleHint;
        wrapper.setState({ show: true });
        expect(wrapper.state('show')).toBe(true);
        instance.closePopup();
        expect(wrapper.state('show')).toBe(false);
    });

    it('it checks click outside', () => {
        const wrapper = shallow(<BubbleHint />);
        const instance = wrapper.instance() as BubbleHint;

        wrapper.setState({ show: true });
        expect(wrapper.state('show')).toBe(true);
        instance.getContainer(div);
        instance.handleClickOutside({ target: {} } as React.SyntheticEvent<EventTarget>);
        expect(wrapper.state('show')).toBe(false);
    });

    it('it checks click outside with closed popup', () => {
        const wrapper = shallow(<BubbleHint />);
        const instance = wrapper.instance() as BubbleHint;

        instance.getContainer(div);
        instance.handleClickOutside({ target: {} } as React.SyntheticEvent<EventTarget>);
        expect(wrapper.state('show')).toBe(false);
    });

    it('it handles handle click', () => {
        const onClick = jest.fn();
        const wrapper = shallow(<BubbleHint onClick={onClick} />);
        const instance = wrapper.instance() as BubbleHint;
        div.contains = () => { return true; };

        instance.getTrigger(div);
        instance.handleClick({ target: {} } as React.MouseEvent<HTMLElement>);
        expect(onClick.mock.calls).toHaveLength(1);
    });

    it('it handles handle click not contains element', () => {
        const onClick = jest.fn();
        const wrapper = shallow(<BubbleHint onClick={onClick} />);
        const instance = wrapper.instance() as BubbleHint;

        instance.getTrigger(div);
        instance.handleClick({ target: {} } as React.MouseEvent<HTMLElement>);
        expect(onClick.mock.calls).toHaveLength(0);
    });

    it('it handles onMouseLeave', () => {
        const onMouseLeave = jest.fn();
        const wrapper = shallow(<BubbleHint onMouseLeave={onMouseLeave} />);
        const instance = wrapper.instance() as BubbleHint;
        instance.handleMouseLeave({} as React.SyntheticEvent<EventTarget>);
        expect(onMouseLeave.mock.calls).toHaveLength(1);
    });

    it('it handles onMouseEnter', () => {
        const onMouseEnter = jest.fn();
        const wrapper = shallow(<BubbleHint onMouseEnter={onMouseEnter} />);
        const instance = wrapper.instance() as BubbleHint;

        instance.handleMouseEnter({} as React.SyntheticEvent<EventTarget>);
        expect(onMouseEnter.mock.calls).toHaveLength(1);
    });

    it('it handles componentWillUnmount', () => {
        const componentWillUnmount = jest.spyOn(BubbleHint.prototype, 'componentWillUnmount');
        const wrapper = shallow(<BubbleHint />);
        wrapper.unmount();
        expect(componentWillUnmount.mock.calls).toHaveLength(1);
    });
});
