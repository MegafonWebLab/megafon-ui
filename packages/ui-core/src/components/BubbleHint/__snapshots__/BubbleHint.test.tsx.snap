// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BubbleHint /> it renders BubbleHint 1`] = `
<Manager
  className="mfui-bubble-hint"
  tag="div"
>
  <div
    className="mfui-bubble-hint__container"
    onClick={[Function]}
  >
    <Target
      className="mfui-bubble-hint__target"
    >
      <div
        className="mfui-bubble-hint__trigger"
      />
    </Target>
    <Popper
      className="mfui-bubble-hint__popup mfui-bubble-hint__popup_width_normal"
      component="div"
      eventsEnabled={false}
      modifiers={
        Object {
          "flip": Object {
            "behavior": Array [
              "right",
              "left",
              "bottom",
              "top",
            ],
          },
        }
      }
      placement="top"
      positionFixed={false}
    >
      <div
        className="mfui-bubble-hint__popup-inner mfui-bubble-hint__popup-inner_padding_normal mfui-bubble-hint__popup-inner_align_left"
      >
        <div
          className="mfui-bubble-hint__content"
        />
        <Arrow
          className="mfui-bubble-hint__arrow"
        />
      </div>
    </Popper>
  </div>
</Manager>
`;

exports[`<BubbleHint /> it renders with click 1`] = `
<Manager
  className="mfui-bubble-hint"
  tag="div"
>
  <div
    className="mfui-bubble-hint__container"
    onClick={[Function]}
  >
    <Target
      className="mfui-bubble-hint__target"
    >
      <div
        className="mfui-bubble-hint__trigger"
      />
    </Target>
    <Popper
      className="mfui-bubble-hint__popup mfui-bubble-hint__popup_width_normal"
      component="div"
      eventsEnabled={false}
      modifiers={
        Object {
          "flip": Object {
            "behavior": Array [
              "right",
              "left",
              "bottom",
              "top",
            ],
          },
        }
      }
      placement="top"
      positionFixed={false}
    >
      <div
        className="mfui-bubble-hint__popup-inner mfui-bubble-hint__popup-inner_padding_normal mfui-bubble-hint__popup-inner_align_left"
      >
        <div
          className="mfui-bubble-hint__content"
        />
        <Arrow
          className="mfui-bubble-hint__arrow"
        />
      </div>
    </Popper>
  </div>
</Manager>
`;
